tests/requests/glacier/tree_hash_tests.rb

```diff
@@ -59,4 +59,26 @@
 
   end
 
+  tests('multipart with unaligned parts') do
```

Comment:
It took me pondering this for a bit to understand what you meant by unaligned. So I think it might be good (for me and for posterity) if we were more explicit. If I understand, the issue is "multipart with parts less than one megabyte" that causes the issue. My initial guess was that having some x's in the first part and the rest in the second part was the issue, but if I'm not mistaken, it would also fail if the first 512*1024 were 'a' and the second were 'b', in the same way. Does that make sense? Is my understanding accurate?

Reply:
Unaligned means not starting on (one megabyte) boundary, that is not starting on offset which is multiple of one megabyte. Obviously, only way for some parts to not start on one megabyte boundary is when some previous part is not ending on one, but if it's **last** part not ending on the boundary the original code would already work.

It's true that if first part would be 512*1024*'a' and then you would add second part with some b's it would be same problem as the b's would be unaligned. However, you can reproduce the problem also with all parts bigger than one megabyte - if the first part would be 1.5MB (second is 2MB and third is 1.5MB), all parts would be bigger than one megabyte but only the first would be aligned because neither 1.5MB nor 3.5MB is multiple of 1MB.

Note that one megabyte is only the **smallest** size exhibiting the problem. Original code would also hit problem if you would add a 2MB part not aligned on 2MB (not starting on 2MB boundary).

It's the same sense of "align" as described on https://en.wikipedia.org/wiki/Data_structure_alignment - except we are not dealing with data in memory, but with parts in "virtual" space of whole file.

As someone who started programming on C and assembler, this is obvious to me, but it's true you don't deal with it so often in higher languages so maybe it's not so clear nowadays?

Reply:
Now, regarding the code: the most important part to understand is the "stack" array. Stack is again term from low level programming, but that's not important here. Important is that first position in that array corresponds with 1MB, second 2MB, third 4MB, fourth 8MB and so on. In any time, the size of all already added parts is equal to sum of all existing (non-nil) positions multiplied by that number, plus last_chunk_length for the remainder smaller than one megabyte. So, if last_chunk_length is half megabyte, stack[0] is filled, stack[1] and stack[2] empty and stack[3] filled, the size is 0.5MB + 1*1MB + 0*2MB + 0*4MB + 0*8MB = 9.5MB.

I guess this part would **definitely** warrant explanation comment in code.

Reply:
Thanks for the extra explanation. Yeah, I think if you could add some explanation comments that might be helpful. I'm pretty sure I now understand, but I'm worried others might not (and/or when I come back to it later I won't recall). Thanks!